Add `--file` flag for `docker secret create` command

yongtang opened the pull request:
**- What I did**

This fix tries to address the issue raised in #28581 and #28927 where it is not possible to create a secret from a file (only through STDIN).

**- How I did it**

This fix add a flag `--file` to `docker secret create` so that it is possible to create a secret from a file with:
```
docker secret create --file secret.in secret.name
```

or
```
echo TEST | docker secret create --file - secret.name
```

Related docs has been updated.

**- How to verify it**

An integration test has been added to cover the changes.

**- Description for the changelog**

Add `--file` flag for `docker secret create`.

**- A picture of a cute animal (not mandatory but encouraged)**

![cute-lion-cubs-animals-pictures-pics](https://cloud.githubusercontent.com/assets/6932348/20459716/7d392382-ae80-11e6-9d97-6f28196840b0.jpg)


This fix fixes #28581.
This fix is related to #28927.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add `--file` flag for `docker secret create` command

This fix tries to address the issue raised in 28581 and 28927
where it is not possible to create a secret from a file (only
through STDIN).

This fix add a flag `--file` to `docker secret create` so that
it is possible to create a secret from a file with:
```
docker secret create --file secret.in secret.name
```

or
```
echo TEST | docker secret create --file - secret.name
```

Related docs has been updated.

An integration test has been added to cover the changes.

This fix fixes 28581.
This fix is related to 28927.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

vdemeester commented:
/cc @thaJeztah @ehazlett

thaJeztah commented:
I think we should have;

- an `-` option on `docker create secret` (like we have on `docker build` to specify that we want to use `stdin` as input 
- have a `--file` flag

but @diogomonica for thoughts as well

diogomonica commented:
Given that I think it's trivial for people to do `cat file | docker secret create`, I don't think we need a file option, and having to specify --input is annoying, so I would say we don't change anything.

This said, I'm fine with whatever makes it more consistent with the rest of docker. 

My requirements are that you can do `openssl rand | docker secret create`

thaJeztah commented:
> Given that I think it's trivial for people to do `cat file | docker secret create`, I don't think we need a file option

Perhaps we should check if that's just as easy / common on Windows.

> This said, I'm fine with whatever makes it more consistent with the rest of docker.

I did a quick check on what we have right now. On `docker load`, we have an `--input` flag. For `docker import`, and `docker build`, the file is the first argument (so `-` has to be specified to load from `stdin`).

One improvement we _can_ make is to add better handling if no input is provided. It's confusing that when running `docker secret create`, it seems to "hang". For `docker load` we added a check, and it prints a message if no input was provided;

```bash
docker load
requested load from stdin, but stdin is empty
```

cpuguy83 commented:
Typically CLI tools that support input via stdin also have a `--file` (or whatever), so SGTM.

> Perhaps we should check if that's just as easy / common on Windows.
Don't see why it wouldn't work, IIRC shell pipes are supported on Powershell.
@jhowardmsft

thaJeztah commented:
Opened https://github.com/docker/docker/pull/28661 to improve  the feedback if no input was provided

ehazlett commented:
I'm also SGTM for `--file`.  @thaJeztah +1 for error on no feedback.

yongtang commented:
This issue is also related to #28927. From the comment (https://github.com/docker/docker/pull/28629#issuecomment-261963379) in this PR and in #28927, it seems that the consensus is to introduce a `--file [file|-]` flag to explicit specify the input.

I updated the PR to reflect the new changes. The PR description and title has been changed as well.

yongtang commented:
/cc @aaronlehmann @aluzzardi @diogomonica @cyli @tianon please take a look as well.

ehazlett commented:
LGTM (NAM)

aaronlehmann commented on cli/command/secret/create.go:
```diff
@@ -43,9 +46,29 @@ func runSecretCreate(dockerCli *command.DockerCli, options createOptions) error
+	if options.file == "-" {
+		// To avoid getting stuck, verify that a file is given in STDIN
+		if dockerCli.In().IsTerminal() {
+			return fmt.Errorf("Requested from STDIN ('-'), but STDIN is empty")
```
I think it should be okay to type in the secret on stdin. I don't know of any other unix tools that disallow this. And the error message doesn't seem right for this case.

yongtang replied:
Thanks @aaronlehmann. The PR has been updated with the restriction removed.

aaronlehmann commented:
LGTM

diogomonica commented:
LGTM

aluzzardi commented:
Does this also work with:

```
docker secret create name -f file
```

That is, having flags after the secret name?

yongtang commented:
Thanks @aluzzardi. Yes the `docker secret create name -f file` works as well. I updated the PR and added a test case to the integration test to cover it. Please take a look.

ehazlett commented:
@yongtang would you mind rebasing?  thx

yongtang commented:
Thanks @ehazlett. The PR has been rebased.

ehazlett commented:
thanks!

On Fri, Dec 9, 2016 at 5:31 PM, Yong Tang <notifications@github.com> wrote:

> Thanks @ehazlett <https://github.com/ehazlett>. The PR has been rebased.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/28629#issuecomment-266072171>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAP6Isy5aCMAlsLWJKoM-TbDpq4dJ8Dxks5rGZBdgaJpZM4K3bQu>
> .
>

thaJeztah commented:
@yongtang didn't this PR have a check if stdin is empty (like in https://github.com/docker/docker/pull/28661)? I seem to recall doing this;

```
docker secret create changes -f -
```

produced an error

otherwise LGTM

yongtang commented:
@thaJeztah There were some concerns about producing an error in that situation so the check was removed later (see https://github.com/docker/docker/pull/28629#discussion_r90507453)

thaJeztah reviewed (approved):
@yongtang a, I see, I overlooked that (just recalled testing this a while back). Given that it's now an explicit choice to use `stdin`, I think that's ok

LGTM

vdemeester commented on cli/command/secret/create.go:
```diff
@@ -43,9 +46,23 @@ func runSecretCreate(dockerCli *command.DockerCli, options createOptions) error
 
-	secretData, err := ioutil.ReadAll(os.Stdin)
+	if options.file == "" {
+		return fmt.Errorf("Please specify either a file name or STDIN ('-')  with --file")
```
nit, but there is one space too much between `('-')` and `with --file` :sweat_smile:

yongtang replied:
Thanks @vdemeester. That is definitely not easy to spot this place 😅 . The PR has been updated, please take a look.

vdemeester commented on integration-cli/docker_cli_secret_create_test.go:
```diff
@@ -104,3 +108,33 @@ func (s *DockerSwarmSuite) TestSecretCreateResolve(c *check.C) {
 	c.Assert(out, checker.Not(checker.Contains), fake)
 }
+
+func (s *DockerSwarmSuite) TestSecretCreateWithFile(c *check.C) {
```
In the end there could be a unit test for that. But I guess there was no *cli integration test* on `docker secret create` so, I'm ok with that :angel:

vdemeester reviewed (changes requested):
Small comments, otherwise looks good

vdemeester reviewed (approved):
LGTM 🐻

aaronlehmann commented:
> > didn't this PR have a check if stdin is empty (like in #28661)?

> There were some concerns about producing an error in that situation so the check was removed later (see https://github.com/docker/docker/pull/28629#discussion_r90507453)

I might be confused, but the comment you linked to seems to be attached to this (removed) code:

```go
	if options.file == "-" {
		// To avoid getting stuck, verify that a file is given in STDIN
		if dockerCli.In().IsTerminal() {
			return fmt.Errorf("Requested from STDIN ('-'), but STDIN is empty")
```

I don't see that checking that stdin is empty, which is why I commented on it ("...And the error message doesn't seem right for this case."). I might be missing some context, though. I would be fine with having a check, but this would have to happen after the data is read, because otherwise how can you know if you receive no data on stdin?

Anyway, there's a check on the swarmkit side, so it doesn't really matter if we check this on the client side.

thaJeztah commented:
@aaronlehmann the error was shown if I tried to create a secret from stdin, but didn't provide any content. Right now, the CLI blocks until content is provided. Now that selecting `stdin` is an explicit action (i.e., I need to set `-f -`), I don't think that's a problem (from a UX perspective)

aaronlehmann commented:
Agreed. The only thing we might consider adding is an error message if you hit `^D` without typing anything. But as I mentioned, such an error should be sent back from the daemon anyway (mind giving it a try?).

thaJeztah commented:
@aaronlehmann Detaching without typing does produce an error, so that looks ok (error presentation could be nicer, but that's for other errors as well);

```
docker secret create foo -f -
^D
Error response from daemon: rpc error: code = 3 desc = secret data must be larger than 0 and less than 512000 bytes
```
